### PR TITLE
BREAKING: Replaces `ComposeScreenViewFactory` with `ScreenComposableFactory`

### LINKED ISSUE #546: Wrapper renderings introduce an extra AndroidView unnecessarily when the wrapped rendering's factory is a ComposeScreenViewFactory.
I just realized this wrapping pattern (even before this change) might introduce an extra view unnecessarily when the wrapped rendering's factory is a `ComposeViewFactory`. It shouldn't affect correctness but something we should probably check/test and to keep in mind as yall adopt compose. 

__Originally posted by @zach-klippenstein in https://github.com/square/workflow-kotlin/pull/544#discussion_r704522274__

I think this could probably be addressed if you came up with a way to generalize the wrapping pattern (as @rjrjr suggested on [slack](https://workflow-community.slack.com/archives/CHTFPR277/p1631109753098600?thread_ts=1631109042.098100&channel=CHTFPR277&message_ts=1631109753.098600)), since `WorkflowRendering`'s `asComposeViewFactory` could use that to do the type check at the leaf.

### PR DESCRIPTION
**Pull request by rjrjr**
**_Please be sure to review `workflow-ui/compose/README.md`, which GitHub helpfully hides by default because it's big._**

_There are two commits in this PR, and it'd be wise to review them in sequence:_

### _`ViewRegistry.Key` allows multiple factory types per rendering type_

Paves the way for parallel Classic and Compose implementations of wrappers like `NamedScreen`, to fix #546. Ironically, breaks Compose support in the process, but that is fixed by the next commit.

### _Replaces `ComposeScreenViewFactory` with `ScreenComposableFactory`_

This commit takes advantage of the new `ViewRegistry.Key` (which allows renderings to be bound to multiple UI factories) to fix a long standing problem where wrapper screens -- things like `NamedScreen` and `EnvironmentScreen` -- used in a Compose context would cause needless calls to `@Composeable fun AndroidView()` and `ComposeView()`.

For example, consider this rendering:

```
BodyAndOverlaysScreen(
  body = SomeComposeScreen(
    EnvironmentScreen(
      SomeOtherComposeScreen
    )
  )
)
```

Before this change, that would create a View hierarchy something like this:

```
BodyAndOverlaysContainer : FrameLayout {
  mChildren[0] = ComposeView {
    // compose land
    SomeComposeScreen.Content {
      AndroidView {
        ComposeView {
          // nested compose land
            SomeOtherComposeScreen.Content()
```

Now it will look this way:

```
BodyAndOverlaysContainer : FrameLayout {
  mChildren[0] = ComposeView {
    // compose land
    SomeComposeScreen.Content {
      SomeOtherComposeScreen.Content()
```

`ScreenComposableFactory` replaces `ComposeScreenViewFactory`, and `ComposeScreen` no longer extends `AndroidScreen`. Compose support is now a first class citizen, instead of a hack bolted on to View support.

Unfortunately, `ViewEnvironment.withComposeInteropSupport()` (see below) now must be called near the root of a Workflow app to enable the seamless Compose > Classic > Compose handling that used to be built in to `WorkflowRendering` and `ComposeScreenViewFactory`. This means that call is required for Compose support for built in rendering types like `BodyAndOverlaysScreen` and `BackStackScreen`, which so far are backed only by classic View implementations.

Other introductions, changes:

- `Screen.toComposableFactory()`, used by `WorkflowRendering()` in the same way that `WorkflowViewStub` uses `Screen.toViewFactory()`

- `ScreenComposableFactoryFinder`, a `ViewEnvironment`-based strategy object used by `Screen.toComposableFactory()` the same way that `Screen.toViewFactory()` uses `ScreenViewFactoryFinder`. The default implementation provides Compose bindings for `NamedScreen` and `EnvironmentScreen`, fixing #546.

- `ScreenViewFactoryFinder.getViewFactoryForRendering()` can now return `null`. A `requireViewFactoryForRendering()` extension is introduced for use when `null` is not acceptable.

- `ViewEnvironment.withComposeInteropSupport()`, which wraps the found `ScreenComposableFactoryFinder` and `ScreenViewFactoryFinder` with implementations that allow Compose contexts to handle renderings bound only to `ScreenViewFactory`, and classic contexts to handle renderings bound only to `ScreenComposableFactory`. Replaces the logic that used to be in the private `ScreenViewFactory.asComposeViewFactory()` extension in `WorkflowRendering()`.

- `Screen.Preview()` is introduced. The existing `Preview()` extension functions were tied to `ScreenViewFactory`, making them much less useful. It is still the case that previews work for non-Compose UI code just fine. Which is pretty cool, really.

Fixes #546

### CODE REVIEW

**Comment by rjrjr** [comment]
- [ ] Use `ViewRegistry.getFactoryFor()` or get rid of it
- [ ] Update kdoc to note that `ScreenViewFactoryFinder` method now can return `null`, and mention new `require...` method

**Comment by rjrjr** [comment]
This is looking really solid. I don't quite have our internal test suite passing yet, but so far everything I've found has been newly exposed bugs in our workflow integration, especially the hellish duct tape supporiting our remaining pre-workflow code. I'm pretty optimistic about shipping this month.

File: workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/ViewRegistry.kt
Comment (rjrjr) [inline workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/ViewRegistry.kt:102]: It is very tempting to rename `Entry` to `UiFactory`. I avoided that in this PR to minimize surface area and migration pain. But maybe we should allow ourselves one more round of renaming before locking everything down?

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (rjrjr) [inline workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:131]: Isn't this redundant with `screenComposableFactory(` above?

File: workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/ViewRegistry.kt
Comment (rjrjr) [reply workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/ViewRegistry.kt:102]: Nah.

File: workflow-ui/compose/src/androidTest/java/com/squareup/workflow1/ui/compose/WorkflowRenderingTest.kt
Comment (rjrjr) [inline workflow-ui/compose/src/androidTest/java/com/squareup/workflow1/ui/compose/WorkflowRenderingTest.kt:169]: Ta da!

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/WorkflowRendering.kt
Comment (rjrjr) [inline workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/WorkflowRendering.kt:156]: This became `ScreenComposableFactoryFinder`

File: workflow-ui/compose/README.md
Comment (rjrjr) [inline workflow-ui/compose/README.md:123]: @ekeitho 👀

File: workflow-ui/compose-tooling/src/main/java/com/squareup/workflow1/ui/compose/tooling/Previews.kt
Comment (ekeitho) [inline workflow-ui/compose-tooling/src/main/java/com/squareup/workflow1/ui/compose/tooling/Previews.kt:90]: whoa, neat.

File: workflow-ui/compose/README.md
Comment (ekeitho) [inline workflow-ui/compose/README.md:6]: design-doc*

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt
Comment (vgonda) [inline workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt:26]: Do we need to answer this before merging?

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt
Comment (vgonda) [inline workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt:31]: Should we have a TODO here to add them?

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ViewEnvironmentWithComposeSupport.kt
Comment (vgonda) [inline workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ViewEnvironmentWithComposeSupport.kt:53]: Drawing attention to this lingering TODO in case it needs attention before merging

**Review by vgonda — APPROVED** [review]

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ViewEnvironmentWithComposeSupport.kt
Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ViewEnvironmentWithComposeSupport.kt:53]: Oh, thanks!

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt
Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt:31]: That work isn't planned, so I'm reluctant to imply it. I'll tighten up the language a bit.

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt
Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactoryFinder.kt:26]: Answer appears to be no, deleting.

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ViewEnvironmentWithComposeSupport.kt
Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ViewEnvironmentWithComposeSupport.kt:53]: Going to hold off on this until the need makes itself clear.

File: workflow-ui/compose-tooling/src/main/java/com/squareup/workflow1/ui/compose/tooling/Previews.kt
Comment (ijwhelan) [inline workflow-ui/compose-tooling/src/main/java/com/squareup/workflow1/ui/compose/tooling/Previews.kt:39]: 😮

**Review by ijwhelan — APPROVED** [review]

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (ekeitho) [inline workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:26]: trying to understand this better, i was thinking this would of been of type `ComposeScreen` to make it more assertive to only be used with compose; but after reading the comment below `Use [ScreenComposableFactory] directly if you need to prevent your [Screen] rendering classes from depending on Compose at compile time.` and the being able to inject things like `NamedScreen`, `EnvironmentScreen` into the compose content - is the reasoning why?

maybe having that sort of comment near the top why it isn't `ComposeScreen` would be helpful for readability.

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (ekeitho) [inline workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:84]: outside of scope of this, but out of curiosity if we were wanting to add a compose overlay factory.

would that entail creating a new `ComposeOverlayDialogFactory` which implements `ScreenComposableFactory` & then adding to the [checks in ScreenComposableFactoryFinder](https://github.com/square/workflow-kotlin/pull/1146/files#diff-488dc3940ee9cf848df3a29160328f6b811dcca43c0f130d52d10d66fad40cfdR30) searching for `ComposeOverlayDialogFactory`?

or would that entail creating a new factory like this one, with it's relative factory finder?

**Review by ekeitho — APPROVED** [review]

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:26]: `ComposeScreen` is a possible implementation, one of many possible concrete implementations of `Screen`. Tying this to `ComposeScreen` would be like tying it to `ArrayList` instead of `List`. Really, `ComposeScreen` is just a convenience so that you don't have to deal with `ScreenComposableFactory` directly.

Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:26]: I'll find a way to get that across better in the docs.

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:84]: That would require:

- A Compose binding (`ScreenComposableFactory` implementation) for `BodyAndOverlaysScreen` -- an alternative to the existing `BodyAndOverlaysContainer`

- A new `OverlayComposableFactory` / `OverlayComposableFactoryFinder` pairing

That's assuming you weren't going to use any non-Compose dialog code at all.

If you were going to try to make it coexist with the `DialogComposableFactory` you would also need to enhance the logic in `withComposeInteropSupport()`. But that would only be feasible if your Compose version was still going to create `Dialog` instances, so that Compose-managed ones could interleave with legacy instances. Even then, you would somehow have to make `DialogCollator` et al agnostic between `Dialog` intances created via `Compose` v. those created the other way. No idea if that's practical.

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (rjrjr) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:26]: @ekeitho I just pushed up a separate SHA that tweaks some kdoc, PTAL.

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (ekeitho) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:85]: yeah exactly not going to use any non-compose dialog code at all - thanks for explaining!

File: workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt
Comment (ekeitho) [reply workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt:26]: this makes a lot more sense to me, thank you!

**Comment by rjrjr** [comment]
As soon as this merges (still trying to get our #*&$(& benchmarks to run), we should follow up with a new Compose-based BackStackScreen treatment that fixes #669.

**Comment by rjrjr** [comment]
> As soon as this merges (still trying to get our #*&$(& benchmarks to run), we should follow up with a new Compose-based BackStackScreen treatment that fixes #669.

Benchmarks look good. No real improvement in the instrumented portion of our real world app, but that's to be expected since our Compose code is still almost entirely at the leaf level. But the reassurance that the latency around those leaves didn't climb is meaningful.

File: workflow-ui/compose/build.gradle.kts
Comment (rjrjr) [inline workflow-ui/compose/build.gradle.kts:16]: Oops, didn't mean to merge this.

Comment (rjrjr) [reply workflow-ui/compose/build.gradle.kts:16]: No, not accident, required by a crummy old AVD version. But this ripples to break other things (how did I get a green PR?). Need to consider spreading the blast radius or else upgrading the AVDs.

@RBusarow @steve-the-edwards, what's your kneejerk?

**Pull request merged**
